feat: Updates to Go 1.21.3

lantoli opened the pull request:
## Description

Jira ticket: [INTMDB-1209](https://jira.mongodb.org/browse/INTMDB-1209)

Updates to Go 1.21.3

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [X] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

## Further comments

lantoli committed:
update to Go 1.21.3

lantoli committed:
update doc

lantoli committed:
update linter

lantoli committed:
keep timeout in linter to avoid error: Timeout exceeded: try increasing it by passing --timeout option

lantoli committed:
adjust linter to use depguard v2

lantoli commented on .golangci.yml:
```diff
@@ -49,11 +49,34 @@ linters-settings:
   maligned:
     suggest-new: true
   depguard:
```
depguard v2 is used in this new linter version and config is totally different

gssbzn replied:
we (CLI, SDKs) decided to disabled the linter since we weren't really using it or had any concerns on a wrong import

lantoli replied:
do you mean depguard or all linters?

gssbzn replied:
`depguard` we removed it from the list of enabled linters

gssbzn replied:
you can run `golangci-lint linters` in the project folder to see a list of enabled/disabled linters

lantoli replied:
ok, i'll do. I see some value, e.g. we're using github.com/go-test/deep out of tests and we probably shouldn't, but I agree to remove it. thx

gssbzn commented on CONTRIBUTING.md:
```diff
@@ -1,25 +1,28 @@
+- Run `make build` to build the binary in the `./bin` directory:
 - Use the local provider binary in the `./bin` folder:
-  - Create the following `dev.trfc` file inside your directory 
+  - Create the following `dev.trfc` file inside your directory
```
```suggestion
 - Create the following `dev.trfc` file inside your directory
```
this file seems like autoformated but since here

lantoli replied:
reformat undone as it's not the purpose of this PR

gssbzn commented on CONTRIBUTING.md:
````diff
@@ -90,87 +99,101 @@ You must also configure the following environment variables before running the t
   ```
-~> **Notice:** For more information about the PrivateLink (for AWS only), see: https://docs.atlas.mongodb.com/reference/api/encryption-at-rest/https://docs.atlas.mongodb.com/reference/api/private-endpoint/
+
+~> **Notice:** For more information about the PrivateLink (for AWS only), see: <https://docs.atlas.mongodb.com/reference/api/encryption-at-rest/https://docs.atlas.mongodb.com/reference/api/private-endpoint/>
````
in case you want to improve how this renders https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

gssbzn commented on CONTRIBUTING.md:
meta comment and not for this PR, this file is huge and should probably be split by topics

lantoli replied:
you're right, thx

lantoli replied:
undoing markdown reformat as it doesn't really belong to this PR

gssbzn commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/mongodb/terraform-provider-mongodbatlas
 
-go 1.20
+go 1.21.3
```
```suggestion
go 1.21
```
if you want to specify a patch use [toolchain](https://go.dev/doc/toolchain) (this is a new feature for 1.21)

anecdotally I just fixed this issue in the Atlas operator https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1188

lantoli replied:
uhm, i removed the line and "go mod tidy" did this for me

lantoli replied:
changed

gssbzn reviewed (commented):
general LGTM except for specifying a patch version in the go version

lantoli committed:
apply feedback about using toolchain version

lantoli committed:
apply feedback to remove depguard linter

lantoli committed:
undo reformat

gssbzn reviewed (approved):
LGTM

lantoli committed:
remove linters not being used because abandoned.

WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused.
WARN [runner] The linter 'structcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused.
WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused.
